mutate/mutate_fuzzer.go

```diff
@@ -0,0 +1,180 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package mutate
+
+import (
+	"archive/tar"
+	"bytes"
+	"context"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	fuzz "github.com/AdaLogics/go-fuzz-headers"
+	"github.com/opencontainers/go-digest"
+	imeta "github.com/opencontainers/image-spec/specs-go"
+	ispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/opencontainers/umoci/oci/cas"
+	casdir "github.com/opencontainers/umoci/oci/cas/dir"
+	"github.com/opencontainers/umoci/oci/casext"
+)
+
+// fuzzSetup() does the necessary setup for the fuzzer
+// It takes a data parameter provided by the fuzzer
+func fuzzSetup(dir string, data []byte) (cas.Engine, ispec.Descriptor) {
+	dir = filepath.Join(dir, "image")
+	if err := casdir.Create(dir); err != nil {
+		panic(err)
+	}
+
+	engine, err := casdir.Open(dir)
+	if err != nil {
+		panic(err)
+	}
+	engineExt := casext.NewEngine(engine)
+
+	// Write a tar layer.
+	var buffer bytes.Buffer
+	tw := tar.NewWriter(&buffer)
+	tw.WriteHeader(&tar.Header{
+		Typeflag: tar.TypeReg,
+		Name:     "test",
+		Mode:     0644,
+		Size:     int64(len(data)),
+	})
+	tw.Write(data)
+	tw.Close()
+
+	// Push the base layer.
+	diffidDigester := cas.BlobAlgorithm.Digester()
+	hashReader := io.TeeReader(&buffer, diffidDigester.Hash())
+	layerDigest, layerSize, err := engine.PutBlob(context.Background(), hashReader)
+	if err != nil {
+		panic(err)
+	}
+
+	// Create a config.
+	config := ispec.Image{
+		Config: ispec.ImageConfig{
+			User: "default:user",
+		},
+		RootFS: ispec.RootFS{
+			Type:    "layers",
+			DiffIDs: []digest.Digest{diffidDigester.Digest()},
+		},
+		History: []ispec.History{
+			{EmptyLayer: false},
+		},
+	}
+
+	configDigest, configSize, err := engineExt.PutBlobJSON(context.Background(), config)
+	if err != nil {
+		panic(err)
+	}
+
+	// Create the manifest.
+	manifest := ispec.Manifest{
+		Versioned: imeta.Versioned{
+			SchemaVersion: 2,
+		},
+		Config: ispec.Descriptor{
+			MediaType: ispec.MediaTypeImageConfig,
+			Digest:    configDigest,
+			Size:      configSize,
+		},
+		Layers: []ispec.Descriptor{
+			{
+				MediaType: ispec.MediaTypeImageLayerGzip,
+				Digest:    layerDigest,
+				Size:      layerSize,
+			},
+		},
+	}
+
+	manifestDigest, manifestSize, err := engineExt.PutBlobJSON(context.Background(), manifest)
+	if err != nil {
+		panic(err)
+	}
+
+	return engine, ispec.Descriptor{
+		MediaType: ispec.MediaTypeImageManifest,
+		Digest:    manifestDigest,
+		Size:      manifestSize,
+	}
+}
+
+// FuzzMutate implements the fuzzer
+func FuzzMutate(data []byte) int {
+	c := fuzz.NewConsumer(data)
+	byteArray, err := c.GetBytes()
+	if err != nil {
+		return -1
+	}
+	dir, err := ioutil.TempDir("", "umoci-TestMutateAdd")
+	if err != nil {
+		return -1
```

Comment:
Maybe I missed it, but what do -1, 0, 1, and panic() signify here? Seems like these -1s should be panic()s maybe, and 0 and 1 are fuzzer failure/success codes?

Reply:
From my understanding there are 4 possible results you get from a fuzzed function:

 * If it panics, that indicates a crash (bug) and will be kept as a "bad" sample.
 * If it returns `-1`, there was some other error than an input-related crash.
 * If it returns `0`, the input was invalid.
 * If it returns `1`,  the input was valid.

You don't want to `panic()` from a `Fuzz...` function because that means any input which happens to coincide with an unrelated error will be treated as a "bad" sample.

Reply:
Regarding the return values, the relevant documentation can be found here: https://github.com/dvyukov/go-fuzz#usage

However, [the significance of these is low](https://github.com/tailscale/tailscale/pull/1115#discussion_r556171316) and using different return values is more a matter of code structure and formatting than performance, ie: returning `-1` during set up and initialization (`if err!=nil`), returning `0` when fuzzing the target (`if err!=nil`), and returning `1` if the fuzzers succeeds.

The fuzzing engine does not rely on the return values to determine if the input was valid.

Regarding the panics: You don't want panics in a fuzzer, if these are fuzz blockers that prevent it from running continuously. In our case, we build the fuzzers in libfuzzer mode, and panics will stop the fuzzer, so they will be fuzz blockers. However, I would prefer to have the panics in the fuzzer for now, as it will be good to know if the fuzzer fails during initialization, and panics will provide this feedback whereas returning `-1` will not.

Reply:
> 

Regarding the panics: You don't want panics in a fuzzer, if these are fuzz blockers that prevent it from running continuously. In our case, we build the fuzzers in libfuzzer mode, and panics will stop the fuzzer, so they will be fuzz blockers. However, I would prefer to have the panics in the fuzzer for now, as it will be good to know if the fuzzer fails during initialization

But isn't the goal to find panic()s during initialization? For example, you want to find invalid array accesses from PutBlobJSON() or whatever; panic()ing on an error from that is confusing to me. Seems like it should be a `return 0` as well? Or is the goal to move it to `return 0` in the near future? (If so, can you put that in the commit message?)

Reply:
> is the goal to move it to return 0 in the near future?

This was the plan, but I have changed all panics to return 0's now. I am marking this one as resolved, but feel free to reopen if you think otherwise.